li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/largemessage/MessageAssemblerImpl.java

```diff
@@ -45,22 +49,34 @@ public AssembleResult assemble(TopicPartition tp, long offset, byte[] segmentByt
     }
 
     LargeMessageSegment segment = _segmentDeserializer.deserialize(tp.topic(), segmentBytes);
-    if (segment == null) {
-      //not a segment
-      return new AssembleResult(segmentBytes, offset, offset);
-    } else {
-      //sanity-check the segment
-      segment.sanityCheck();
+    return assembleSegment(tp, offset, segmentBytes, segment);
+  }
 
-      // Return immediately if it is a single segment message.
-      if (segment.numberOfSegments == 1) {
-        return new AssembleResult(segment.payloadArray(), offset, offset);
-      } else {
-        LargeMessage.SegmentAddResult result = _messagePool.tryCompleteMessage(tp, offset, segment);
-        return new AssembleResult(result.serializedMessage() == null ? INCOMPLETE_RESULT : result.serializedMessage(),
-                                  result.startingOffset(), offset);
-      }
+  @Override
+  public AssembleResult assemble(TopicPartition tp,
+                                 long offset,
+                                 byte[] segmentBytes,
+                                 Header header) {
+    if (segmentBytes == null) {
+      return new AssembleResult(null, offset, offset);
+    }
+
+    // no LM record header or wrong key, use default assemble()
+    if (header == null || !header.key().equals(Constants.LARGE_MESSAGE_HEADER)) {
+      return assemble(tp, offset, segmentBytes);
+    }
+    // retrieve segment header
+    LargeMessageHeaderValue segmentHeader = LargeMessageHeaderValue.fromBytes(header.value());
+    LargeMessageSegment segment = null;
+    ByteBuffer byteBuffer = ByteBuffer.wrap(segmentBytes);
+    // skip payload header
+    byteBuffer.position(PAYLOAD_HEADER_LENGTH);
```

Comment:
I don't understand why this is being done here. This `assembler` implementation should only be invoked when the large message header exists **AND** its version is `>= V3`.  So, there should be no segment header in the payload byte array anyway.

Reply:
I think you should unit tests for this scenario.

Reply:
1. Sure. Your point is correct and I have added the check to get over this case.

2. Will add.